tests/testthat/test-c.R

```diff
@@ -272,24 +272,43 @@ test_that("vec_c() falls back to c() if S3 method is available", {
   )
 })
 
-test_that("c() fallback is consistent (FIXME)", {
+test_that("c() fallback is consistent", {
+  dispatched <- function(x) structure(x, class = "dispatched")
+  c_method <- function(...) dispatched(NextMethod())
+
   out <- with_methods(
-    c.vctrs_foobar = function(...) structure(NextMethod(), class = "dispatched"),
+    c.vctrs_foobar = c_method,
     list(
       direct = vec_c(foobar(1L), foobar(2L)),
       df = vec_c(data_frame(x = foobar(1L)), data_frame(x = foobar(2L))),
       tib = vec_c(tibble(x = foobar(1L)), tibble(x = foobar(2L))),
       foreign_df = vec_c(foobaz(data_frame(x = foobar(1L))), foobaz(data_frame(x = foobar(2L))))
     )
   )
+  expect_equal(out$direct, dispatched(1:2))
+  expect_equal(out$df$x, dispatched(1:2))
+  expect_equal(out$tib$x, dispatched(1:2))
+  expect_equal(out$foreign_df$x, dispatched(1:2))
+
+  # Hard case: generic record vectors
+  my_rec_record <- function(x) {
+    new_rcrd(list(x = x), class = "my_rec_record")
+  }
 
-  # Proper `c()` dispatch:
-  expect_identical(out$direct, structure(1:2, class = "dispatched"))
-
-  # Inconsistent:
-  expect_identical(out$df$x, foobar(1:2))
-  expect_identical(out$tib$x, foobar(1:2))
-  expect_identical(out$foreign_df$x, foobar(1:2))
+  out <- with_methods(
+    c.vctrs_foobar = c_method,
+    vec_ptype2.my_rec_record.my_rec_record = function(x, y, ...) {
+      my_rec_record(vec_ptype2(field(x, "x"), field(y, "x"), ...))
+    },
+    vec_cast.my_rec_record.my_rec_record = function(x, to, ...) {
+      x
```

Comment:
This should probably cast `field(x, "x")` to `field(to, "x")`, otherwise `vec_c(my_rec_record(foobar(1L)), my_rec_record(foobar(2)))` would probably fail?

I know it is a test case, it was just something that I noticed